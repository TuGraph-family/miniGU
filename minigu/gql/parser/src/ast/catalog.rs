//! AST definitions for *catalog-modifying statements*.

use super::{
    CallProcedureStatement, CatalogObjectRef, GraphElementType, GraphExpr, GraphTypeRef, SchemaPath,
};
use crate::macros::base;
use crate::span::{OptSpanned, Spanned, VecSpanned};

pub type LinearCatalogModifyingStatement = VecSpanned<CatalogModifyingStatement>;

#[apply(base)]
pub enum CatalogModifyingStatement {
    Call(CallProcedureStatement),
    CreateSchema(CreateSchemaStatement),
    DropSchema(DropSchemaStatement),
    CreateGraph(CreateGraphStatement),
    DropGraph(DropGraphStatement),
    CreateGraphType(CreateGraphTypeStatement),
    DropGraphType(DropGraphTypeStatement),
}

#[apply(base)]
pub struct CreateSchemaStatement {
    pub path: Spanned<SchemaPath>,
    pub if_not_exists: bool,
}

#[apply(base)]
pub struct DropSchemaStatement {
    pub path: Spanned<SchemaPath>,
    pub if_exists: bool,
}

#[apply(base)]
pub struct CreateGraphStatement {
    pub path: Spanned<CatalogObjectRef>,
    pub kind: Spanned<CreateGraphOrGraphTypeStatementKind>,
    pub graph_type: Spanned<OfGraphType>,
    pub source: OptSpanned<GraphExpr>,
}

#[apply(base)]
pub enum CreateGraphOrGraphTypeStatementKind {
    Create,
    CreateIfNotExists,
    CreateOrReplace,
}

#[apply(base)]
pub enum OfGraphType {
    Like(Spanned<GraphExpr>),
    Ref(Spanned<GraphTypeRef>),
    Nested(VecSpanned<GraphElementType>),
    Any,
}

#[apply(base)]
pub struct DropGraphStatement {
    pub path: Spanned<CatalogObjectRef>,
    pub if_exists: bool,
}

#[apply(base)]
pub struct DropGraphTypeStatement {
    pub path: Spanned<CatalogObjectRef>,
    pub if_exists: bool,
}

#[apply(base)]
pub struct CreateGraphTypeStatement {
    pub path: Spanned<CatalogObjectRef>,
    pub kind: Spanned<CreateGraphOrGraphTypeStatementKind>,
    pub source: Spanned<GraphTypeSource>,
}

#[apply(base)]
pub enum GraphTypeSource {
    Copy(Spanned<GraphTypeRef>),
    Like(Spanned<GraphExpr>),
    Nested(VecSpanned<GraphElementType>),
}
