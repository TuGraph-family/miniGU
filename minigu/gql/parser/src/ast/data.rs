//! AST definitions for *data-modifying statements*.

use crate::macros::base;

#[apply(base)]
pub struct LinearDataModifyingStatement {}
