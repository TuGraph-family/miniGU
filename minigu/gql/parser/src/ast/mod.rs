pub mod catalog;
pub mod common;
pub mod data;
pub mod lexical;
pub mod object_expr;
pub mod object_ref;
pub mod predicate;
pub mod procedure_call;
pub mod procedure_spec;
pub mod program;
pub mod query;
pub mod session;
pub mod transaction;
pub mod type_element;
pub mod value_expr;
pub mod variable;

pub use catalog::*;
pub use common::*;
pub use data::*;
pub use lexical::*;
pub use object_expr::*;
pub use object_ref::*;
pub use predicate::*;
pub use procedure_call::*;
pub use procedure_spec::*;
pub use program::*;
pub use query::*;
pub use session::*;
pub use transaction::*;
pub use type_element::*;
pub use value_expr::*;
pub use variable::*;
