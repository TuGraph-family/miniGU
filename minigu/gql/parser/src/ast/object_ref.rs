//! AST definitions for *object references*.

use super::Ident;
use crate::macros::base;
use crate::span::{OptSpanned, VecSpanned};

#[apply(base)]
pub enum PredefinedSchemaRef {
    Home,
    Current,
}

#[apply(base)]
pub enum SchemaRef {
    Absolute(SchemaPath),
    Relative(SchemaPath),
    Predefined(PredefinedSchemaRef),
    Parameter(Ident),
}

pub type SchemaPath = VecSpanned<SchemaPathSegment>;

#[apply(base)]
pub enum SchemaPathSegment {
    /// A named object (schema or directory), e.g., `a`.
    Name(Ident),
    /// Parent directory, e.g., `..`.
    Parent,
}

#[apply(base)]
pub enum GraphRef {
    Name(Ident),
    Parameter(Ident),
    Ref(CatalogObjectRef),
    Home,
}

#[apply(base)]
pub enum ProcedureRef {
    Ref(CatalogObjectRef),
    Parameter(Ident),
}

#[apply(base)]
pub enum GraphTypeRef {
    Ref(CatalogObjectRef),
    Parameter(Ident),
}

#[apply(base)]
pub enum BindingTableRef {
    Name(Ident),
    Ref(CatalogObjectRef),
    Parameter(Ident),
}

#[apply(base)]
pub struct CatalogObjectRef {
    pub schema: OptSpanned<SchemaRef>,
    pub objects: VecSpanned<Ident>,
}
