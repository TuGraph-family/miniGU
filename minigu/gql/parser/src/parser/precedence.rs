use core::marker::PhantomData;

use winnow::Parser;
use winnow::combinator::opt;
use winnow::error::ParserError;
use winnow::stream::{Stream, StreamIsPartial};

/// Specifies the associativity of an (infix) operator.
#[derive(Debug, Clone, Copy)]
pub(super) enum Assoc {
    Left,
    Right,
    Neither,
}

pub(super) type Precedence = i64;

/// An operator precedence parser based on the [Pratt algorithm](https://en.wikipedia.org/wiki/Operator-precedence_parser).
/// The implementation is adapted from https://github.com/winnow-rs/winnow/pull/614.
#[allow(clippy::too_many_arguments)]
pub(super) fn precedence<
    I,
    E,
    Operand,
    PrefixOp,
    SuffixOp,
    InfixOp,
    ParseOperand,
    ParsePrefixOp,
    ParseSuffixOp,
    ParseInfixOp,
    FoldPrefix,
    FoldSuffix,
    FoldInfix,
>(
    init: Precedence,
    parse_operand: ParseOperand,
    parse_prefix: ParsePrefixOp,
    parse_suffix: ParseSuffixOp,
    parse_infix: ParseInfixOp,
    fold_prefix: FoldPrefix,
    fold_suffix: FoldSuffix,
    fold_infix: FoldInfix,
) -> impl Parser<I, Operand, E>
where
    I: Stream + StreamIsPartial,
    ParseOperand: Parser<I, Operand, E>,
    ParsePrefixOp: Parser<I, (Precedence, PrefixOp), E>,
    ParseSuffixOp: Parser<I, (Precedence, SuffixOp), E>,
    ParseInfixOp: Parser<I, (Assoc, Precedence, InfixOp), E>,
    FoldPrefix: FnMut(PrefixOp, Operand) -> Result<Operand, E>,
    FoldSuffix: FnMut(Operand, SuffixOp) -> Result<Operand, E>,
    FoldInfix: FnMut(Operand, InfixOp, Operand) -> Result<Operand, E>,
    E: ParserError<I>,
{
    let mut config = PrattConfig {
        parse_operand,
        parse_prefix,
        parse_suffix,
        parse_infix,
        fold_prefix,
        fold_suffix,
        fold_infix,
        _marker: PhantomData,
    };
    move |i: &mut _| config.precedence_impl(i, init)
}

struct PrattConfig<
    I,
    E,
    Operand,
    PrefixOp,
    SuffixOp,
    InfixOp,
    ParseOperand,
    ParsePrefixOp,
    ParseSuffixOp,
    ParseInfixOp,
    FoldPrefix,
    FoldSuffix,
    FoldInfix,
> {
    parse_operand: ParseOperand,
    parse_prefix: ParsePrefixOp,
    parse_suffix: ParseSuffixOp,
    parse_infix: ParseInfixOp,
    fold_prefix: FoldPrefix,
    fold_suffix: FoldSuffix,
    fold_infix: FoldInfix,
    _marker: PhantomData<(I, E, Operand, PrefixOp, SuffixOp, InfixOp)>,
}

impl<
    I,
    E,
    Operand,
    PrefixOp,
    SuffixOp,
    InfixOp,
    ParseOperand,
    ParsePrefixOp,
    ParseSuffixOp,
    ParseInfixOp,
    FoldPrefix,
    FoldSuffix,
    FoldInfix,
>
    PrattConfig<
        I,
        E,
        Operand,
        PrefixOp,
        SuffixOp,
        InfixOp,
        ParseOperand,
        ParsePrefixOp,
        ParseSuffixOp,
        ParseInfixOp,
        FoldPrefix,
        FoldSuffix,
        FoldInfix,
    >
where
    I: Stream + StreamIsPartial,
    ParseOperand: Parser<I, Operand, E>,
    ParsePrefixOp: Parser<I, (Precedence, PrefixOp), E>,
    ParseSuffixOp: Parser<I, (Precedence, SuffixOp), E>,
    ParseInfixOp: Parser<I, (Assoc, Precedence, InfixOp), E>,
    FoldPrefix: FnMut(PrefixOp, Operand) -> Result<Operand, E>,
    FoldSuffix: FnMut(Operand, SuffixOp) -> Result<Operand, E>,
    FoldInfix: FnMut(Operand, InfixOp, Operand) -> Result<Operand, E>,
    E: ParserError<I>,
{
    fn precedence_impl(&mut self, i: &mut I, min_power: Precedence) -> Result<Operand, E> {
        let operand = opt(self.parse_operand.by_ref()).parse_next(i)?;
        let mut operand = if let Some(operand) = operand {
            operand
        } else {
            let len = i.eof_offset();
            let (power, prefix_op) = self.parse_prefix.parse_next(i)?;
            if i.eof_offset() == len {
                return Err(E::assert(i, "`prefix` parsers must always consume"));
            }
            let operand = self.precedence_impl(i, power)?;
            (self.fold_prefix)(prefix_op, operand)?
        };

        let mut prev_op_is_neither = None;
        while i.eof_offset() > 0 {
            let start = i.checkpoint();
            if let Some((power, suffix_op)) = opt(self.parse_suffix.by_ref()).parse_next(i)? {
                if power < min_power {
                    i.reset(&start);
                    break;
                }
                operand = (self.fold_suffix)(operand, suffix_op)?;
                continue;
            }

            let start = i.checkpoint();
            let parse_result = opt(self.parse_infix.by_ref()).parse_next(i)?;
            if let Some((assoc, power, infix_op)) = parse_result {
                let mut is_neither = None;
                let (lpower, rpower) = match assoc {
                    Assoc::Left => (power, power + 1),
                    Assoc::Right => (power, power - 1),
                    Assoc::Neither => {
                        is_neither = Some(power);
                        (power, power + 1)
                    }
                };
                if lpower < min_power || prev_op_is_neither.is_some_and(|p| lpower == p) {
                    i.reset(&start);
                    break;
                }
                prev_op_is_neither = is_neither;
                let rhs = self.precedence_impl(i, rpower)?;
                operand = (self.fold_infix)(operand, infix_op, rhs)?;
                continue;
            }
            break;
        }
        Ok(operand)
    }
}

#[cfg(test)]
mod tests {
    use core::str::FromStr;

    use winnow::ModalResult;
    use winnow::ascii::{digit1, multispace0};
    use winnow::combinator::{delimited, dispatch, empty, fail, peek};
    use winnow::token::any;

    use super::*;

    fn operand(i: &mut &str) -> ModalResult<i64> {
        digit1.try_map(i64::from_str).parse_next(i)
    }

    fn parenthesized(i: &mut &str) -> ModalResult<i64> {
        delimited('(', expr, ')').parse_next(i)
    }

    fn expr(i: &mut &str) -> ModalResult<i64> {
        precedence(
            0,
            delimited(
                multispace0,
                dispatch! {peek(any);
                    '(' => parenthesized,
                    '0'..='9' => operand,
                    _ => fail
                },
                multispace0,
            ),
            delimited(
                multispace0,
                dispatch! {any;
                    '+' => empty.value((2, '+')),
                    '-' => empty.value((2, '-')),
                    _ => fail
                },
                multispace0,
            ),
            delimited(
                multispace0,
                dispatch! {any;
                    '!' => empty.value((3, '!')),
                    _ => fail
                },
                multispace0,
            ),
            delimited(
                multispace0,
                dispatch! {any;
                    '+' => empty.value((Assoc::Left, 0, '+')),
                    '-' => empty.value((Assoc::Left, 0, '-')),
                    '*' => empty.value((Assoc::Left, 1, '*')),
                    '/' => empty.value((Assoc::Left, 1, '/')),
                    _ => fail
                },
                multispace0,
            ),
            |op, a| match op {
                '+' => Ok(a),
                '-' => Ok(-a),
                _ => unreachable!(),
            },
            |a, op| match op {
                '!' => Ok((1..=a).product()),
                _ => unreachable!(),
            },
            |a, op, b| match op {
                '+' => Ok(a + b),
                '-' => Ok(a - b),
                '*' => Ok(a * b),
                '/' => Ok(a / b),
                _ => unreachable!(),
            },
        )
        .parse_next(i)
    }

    #[test]
    fn test_parse_operand() {
        assert_eq!(operand.parse("123").unwrap(), 123);
    }

    #[test]
    fn test_simple_expr() {
        assert_eq!(expr.parse("1 +  3* 4 /2 ").unwrap(), 7);
    }

    #[test]
    fn test_complex_expr() {
        assert_eq!(expr.parse("3- (+ 1 * - (3 +4) + 10)!").unwrap(), -3);
    }
}
