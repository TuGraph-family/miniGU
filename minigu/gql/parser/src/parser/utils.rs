use winnow::Parser;
use winnow::stream::{Location, Stream};

use crate::span::Spanned;

/// A helper macro for executing a parser on an input string. **This should only be used in unit
/// tests!**
#[cfg(all(test, feature = "serde"))]
macro_rules! parse {
    ($parser:expr, $input:literal) => {{
        let input = $input;
        $crate::parser::token::tokenize(input)
            .ok()
            .and_then(|tokens| {
                let stream = crate::parser::token::build_token_stream(&tokens, Default::default());
                $parser.parse(stream).ok()
            })
    }};
}

#[cfg(all(test, feature = "serde"))]
pub(super) use parse;

/// A helper trait for wrapping the output in [`Spanned`].
pub(super) trait ToSpanned<I, O, E>: Parser<I, O, E> {
    /// Wraps the output in [`Spanned`].
    #[inline(always)]
    fn spanned(self) -> impl Parser<I, Spanned<O>, E>
    where
        Self: Sized,
        I: Stream + Location,
    {
        self.with_span().map(|(inner, span)| Spanned(inner, span))
    }
}

/// A helper trait for operating on the value wrapped in [`Spanned`].
pub(super) trait SpannedParserExt<I, O, E>: Parser<I, Spanned<O>, E> {
    /// Unwraps the output from [`Spanned`].
    #[inline(always)]
    fn unspanned(self) -> impl Parser<I, O, E>
    where
        Self: Sized,
    {
        self.map(|spanned| spanned.0)
    }

    /// Maps the inner value of the output while keeping the span unchanged.
    #[inline(always)]
    fn map_inner<F, O2>(self, mut f: F) -> impl Parser<I, Spanned<O2>, E>
    where
        Self: Sized,
        F: FnMut(O) -> O2,
    {
        self.map(move |Spanned(inner, span)| Spanned(f(inner), span))
    }

    /// Updates the span of the output while keeping the inner value unchanged.
    #[inline(always)]
    fn update_span(self) -> impl Parser<I, Spanned<O>, E>
    where
        Self: Sized,
        I: Stream + Location,
    {
        self.with_span().map(|(mut output, new_span)| {
            output.1 = new_span;
            output
        })
    }
}

impl<I, O, E, P> ToSpanned<I, O, E> for P where P: Parser<I, O, E> {}

impl<I, O, E, P> SpannedParserExt<I, O, E> for P where P: Parser<I, Spanned<O>, E> {}

/// A helper macro for defining parser aliases.
macro_rules! def_parser_alias {
    ($name:ident, $parser:ident, $output:ty) => {
        #[inline(always)]
        pub fn $name(
            input: &mut $crate::parser::token::TokenStream,
        ) -> winnow::ModalResult<$output> {
            $parser(input)
        }
    };
}

pub(super) use def_parser_alias;
