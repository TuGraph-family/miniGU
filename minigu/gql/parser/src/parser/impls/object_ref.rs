use winnow::combinator::{alt, dispatch, empty, fail, opt, peek, preceded, repeat, terminated};
use winnow::{ModalResult, Parser};

use super::lexical::{
    binding_table_name, delimited_binding_table_name, delimited_graph_name, directory_name,
    graph_name, object_name, schema_name, substituted_parameter_reference,
};
use crate::ast::*;
use crate::lexer::TokenKind;
use crate::parser::token::{TokenStream, any};
use crate::parser::utils::{SpannedParserExt, ToSpanned, def_parser_alias};
use crate::span::Spanned;

pub fn simple_directory_path<const GREEDY: bool>(
    input: &mut TokenStream,
) -> ModalResult<SchemaPath> {
    if GREEDY {
        repeat(
            1..,
            terminated(
                directory_name.map_inner(SchemaPathSegment::Name),
                TokenKind::Solidus,
            ),
        )
        .parse_next(input)
    } else {
        repeat(
            1..,
            terminated(
                directory_name.map_inner(SchemaPathSegment::Name),
                (
                    TokenKind::Solidus,
                    peek((directory_name, TokenKind::Solidus)),
                ),
            ),
        )
        .parse_next(input)
    }
}

pub fn relative_directory_path<const GREEDY: bool>(
    input: &mut TokenStream,
) -> ModalResult<SchemaPath> {
    (
        repeat(
            1..,
            terminated(
                TokenKind::DoublePeriod
                    .value(SchemaPathSegment::Parent)
                    .spanned(),
                TokenKind::Solidus,
            ),
        ),
        opt(simple_directory_path::<GREEDY>),
    )
        .map(|(mut prefix, suffix): (SchemaPath, _)| {
            prefix.extend(suffix.into_iter().flatten());
            prefix
        })
        .parse_next(input)
}

pub fn absolute_directory_path<const GREEDY: bool>(
    input: &mut TokenStream,
) -> ModalResult<SchemaPath> {
    preceded(TokenKind::Solidus, opt(simple_directory_path::<GREEDY>))
        .map(Option::unwrap_or_default)
        .parse_next(input)
}

pub fn predefined_schema_reference(
    input: &mut TokenStream,
) -> ModalResult<Spanned<PredefinedSchemaRef>> {
    dispatch! {any;
        TokenKind::HomeSchema => empty.value(PredefinedSchemaRef::Home),
        TokenKind::CurrentSchema | TokenKind::Period => empty.value(PredefinedSchemaRef::Current),
        _ => fail
    }
    .spanned()
    .parse_next(input)
}

fn merge_directory_path_and_schema_name(
    mut directory: SchemaPath,
    schema: Spanned<Ident>,
) -> SchemaPath {
    directory.push(Spanned(SchemaPathSegment::Name(schema.0), schema.1));
    directory
}

pub fn relative_catalog_schema_reference<const GREEDY: bool>(
    input: &mut TokenStream,
) -> ModalResult<Spanned<SchemaRef>> {
    alt((
        predefined_schema_reference.map_inner(SchemaRef::Predefined),
        (relative_directory_path::<GREEDY>, schema_name)
            .map(|(directory, schema)| {
                SchemaRef::Relative(merge_directory_path_and_schema_name(directory, schema))
            })
            .spanned(),
    ))
    .parse_next(input)
}

pub fn absolute_catalog_schema_reference<const GREEDY: bool>(
    input: &mut TokenStream,
) -> ModalResult<Spanned<SchemaRef>> {
    (absolute_directory_path::<GREEDY>, schema_name)
        .map(|(directory, schema)| {
            SchemaRef::Absolute(merge_directory_path_and_schema_name(directory, schema))
        })
        .spanned()
        .parse_next(input)
}

pub fn catalog_schema_parent_and_name(input: &mut TokenStream) -> ModalResult<Spanned<SchemaPath>> {
    (absolute_directory_path::<true>, schema_name)
        .map(|(directory, schema)| merge_directory_path_and_schema_name(directory, schema))
        .spanned()
        .parse_next(input)
}

pub fn root_schema_reference(input: &mut TokenStream) -> ModalResult<Spanned<SchemaRef>> {
    TokenKind::Solidus
        .value(SchemaPath::new())
        .map(SchemaRef::Absolute)
        .spanned()
        .parse_next(input)
}

pub fn non_root_schema_reference<const GREEDY: bool>(
    input: &mut TokenStream,
) -> ModalResult<Spanned<SchemaRef>> {
    dispatch! {peek(any);
        TokenKind::Solidus => absolute_catalog_schema_reference::<GREEDY>,
        TokenKind::DoublePeriod
        | TokenKind::HomeSchema
        | TokenKind::CurrentSchema
        | TokenKind::Period => relative_catalog_schema_reference::<GREEDY>,
        TokenKind::SubstitutedParameterReference(_) => {
            reference_parameter_specification.map_inner(SchemaRef::Parameter)
        },
        _ => fail,
    }
    .parse_next(input)
}

pub fn schema_reference(input: &mut TokenStream) -> ModalResult<Spanned<SchemaRef>> {
    alt((non_root_schema_reference::<true>, root_schema_reference)).parse_next(input)
}

def_parser_alias!(
    reference_parameter_specification,
    substituted_parameter_reference,
    Spanned<Ident>
);

pub fn catalog_object_parent_reference(input: &mut TokenStream) -> ModalResult<CatalogObjectRef> {
    dispatch! {peek(any);
        TokenKind::Solidus
        | TokenKind::DoublePeriod
        | TokenKind::HomeSchema
        | TokenKind::CurrentSchema
        | TokenKind::Period
        | TokenKind::SubstitutedParameterReference(_) => {
            alt((
                (
                    non_root_schema_reference::<false>,
                    preceded(TokenKind::Solidus, repeat(0.., terminated(object_name, TokenKind::Period)))
                )
                .map(|(schema, objects)| {
                    CatalogObjectRef {
                        schema: Some(schema),
                        objects
                    }
                }),
                (
                    root_schema_reference,
                    repeat(0.., terminated(object_name, TokenKind::Period)),
                )
                .map(|(schema, objects)| {
                    CatalogObjectRef {
                        schema: Some(schema),
                        objects,
                    }
                }),
            ))
        },
        _ => repeat(1.., terminated(object_name, TokenKind::Period))
            .map(|objects| CatalogObjectRef {
                schema: None,
                objects,
            })
    }
    .parse_next(input)
}

pub fn graph_reference(input: &mut TokenStream) -> ModalResult<Spanned<GraphRef>> {
    dispatch! {peek(any);
        TokenKind::HomeGraph => home_graph,
        TokenKind::DoubleQuoted(_) | TokenKind::AccentQuoted(_) => delimited_graph_name.map_inner(GraphRef::Name),
        _ => {
            alt((
                (catalog_object_parent_reference, graph_name).map(|(mut parent, graph)| {
                    parent.objects.push(graph);
                    GraphRef::Ref(parent)
                }).spanned(),
                reference_parameter_specification.map_inner(GraphRef::Parameter)
            ))
        }
    }
    .parse_next(input)
}

pub fn home_graph(input: &mut TokenStream) -> ModalResult<Spanned<GraphRef>> {
    dispatch! {any;
        TokenKind::HomePropertyGraph | TokenKind::HomeGraph => empty.value(GraphRef::Home),
        _ => fail
    }
    .spanned()
    .parse_next(input)
}

pub fn catalog_object_parent_and_name(
    input: &mut TokenStream,
) -> ModalResult<Spanned<CatalogObjectRef>> {
    (opt(catalog_object_parent_reference), object_name)
        .map(|(parent, graph)| {
            if let Some(mut inner) = parent {
                inner.objects.push(graph);
                inner
            } else {
                CatalogObjectRef {
                    schema: None,
                    objects: [graph].into(),
                }
            }
        })
        .spanned()
        .parse_next(input)
}

def_parser_alias!(
    catalog_graph_parent_and_name,
    catalog_object_parent_and_name,
    Spanned<CatalogObjectRef>
);

pub fn graph_type_reference(input: &mut TokenStream) -> ModalResult<Spanned<GraphTypeRef>> {
    alt((
        catalog_graph_type_parent_and_name.map_inner(GraphTypeRef::Ref),
        reference_parameter_specification.map_inner(GraphTypeRef::Parameter),
    ))
    .parse_next(input)
}

def_parser_alias!(
    catalog_graph_type_parent_and_name,
    catalog_object_parent_and_name,
    Spanned<CatalogObjectRef>
);

pub fn binding_table_reference(input: &mut TokenStream) -> ModalResult<Spanned<BindingTableRef>> {
    dispatch! {peek(any);
        TokenKind::DoubleQuoted(_) | TokenKind::AccentQuoted(_) => delimited_binding_table_name.map_inner(BindingTableRef::Name),
        _ => alt((
            (catalog_object_parent_reference, binding_table_name).map(|(mut parent, name)| {
                parent.objects.push(name);
                BindingTableRef::Ref(parent)
            }).spanned(),
            reference_parameter_specification.map_inner(BindingTableRef::Parameter),
        ))
    }
    .parse_next(input)
}

pub fn procedure_reference(input: &mut TokenStream) -> ModalResult<Spanned<ProcedureRef>> {
    alt((
        catalog_procedure_parent_and_name.map_inner(ProcedureRef::Ref),
        reference_parameter_specification.map_inner(ProcedureRef::Parameter),
    ))
    .parse_next(input)
}

def_parser_alias!(
    catalog_procedure_parent_and_name,
    catalog_object_parent_and_name,
    Spanned<CatalogObjectRef>
);

#[cfg(all(test, feature = "serde"))]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;
    use crate::parser::utils::parse;

    #[test]
    fn test_relative_directory_path_1() {
        let parsed = parse!(relative_directory_path::<true>, "../dir/");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_relative_directory_path_2() {
        let parsed = parse!(relative_directory_path::<true>, "../");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_absolute_directory_path_1() {
        let parsed = parse!(absolute_directory_path::<true>, "/dir/");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_absolute_directory_path_2() {
        let parsed = parse!(absolute_directory_path::<true>, "/");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_predefined_schema_reference_1() {
        let parsed = parse!(predefined_schema_reference, ".");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_predefined_schema_reference_2() {
        let parsed = parse!(predefined_schema_reference, "home_schema");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_predefined_schema_reference_3() {
        let parsed = parse!(predefined_schema_reference, "current_schema");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_absolute_catalog_schema_reference_1() {
        let parsed = parse!(absolute_catalog_schema_reference::<true>, "/s");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_absolute_catalog_schema_reference_2() {
        let parsed = parse!(absolute_catalog_schema_reference::<true>, "/d/s");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_relative_catalog_schema_reference_1() {
        let parsed = parse!(relative_catalog_schema_reference::<true>, "../s");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_relative_catalog_schema_reference_2() {
        let parsed = parse!(relative_catalog_schema_reference::<true>, "../../d/s");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_catalog_object_parent_reference_1() {
        let parsed = parse!(catalog_object_parent_reference, "/");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_catalog_object_parent_reference_2() {
        let parsed = parse!(catalog_object_parent_reference, "/a/b.c.");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_catalog_object_parent_reference_3() {
        let parsed = parse!(catalog_object_parent_reference, "a.b.");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_catalog_object_parent_reference_4() {
        let parsed = parse!(catalog_object_parent_reference, "/a/b/");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_catalog_object_parent_reference_5() {
        let parsed = parse!(catalog_object_parent_reference, "/a.b.");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_catalog_object_parent_reference_6() {
        // Invalid input
        let parsed = parse!(catalog_object_parent_reference, "/a b");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_catalog_object_parent_reference_7() {
        // Invalid input
        let parsed = parse!(catalog_object_parent_reference, "//a.");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_reference_1() {
        let parsed = parse!(graph_reference, "home_graph");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_reference_2() {
        let parsed = parse!(graph_reference, "`graph`");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_reference_3() {
        let parsed = parse!(graph_reference, "$$abc");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_reference_4() {
        let parsed = parse!(graph_reference, "/a/b");
        assert_yaml_snapshot!(parsed);
    }
}
