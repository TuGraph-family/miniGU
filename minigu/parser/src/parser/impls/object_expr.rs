use winnow::combinator::{alt, dispatch, empty, fail, preceded};
use winnow::{ModalResult, Parser};

use super::lexical::object_name_or_binding_variable;
use super::object_ref::graph_reference;
use super::value_expr::{
    non_parenthesized_value_expression_primary_special_case, parenthesized_value_expression,
    value_expression_primary,
};
use crate::ast::{GraphExpr, ObjectExpr};
use crate::lexer::TokenKind;
use crate::parser::token::TokenStream;
use crate::parser::utils::{SpannedParserExt, ToSpanned, peek1, take1};
use crate::span::Spanned;

pub fn graph_expression(input: &mut TokenStream) -> ModalResult<Spanned<GraphExpr>> {
    dispatch! {peek1;
        TokenKind::CurrentPropertyGraph | TokenKind::CurrentGraph => current_graph,
        _ => alt((
            graph_reference.map_inner(GraphExpr::Ref),
            object_expression_primary.map_inner(GraphExpr::Object),
            object_name_or_binding_variable.map_inner(GraphExpr::Name),
            fail,
        ))
    }
    .parse_next(input)
}

pub fn current_graph(input: &mut TokenStream) -> ModalResult<Spanned<GraphExpr>> {
    dispatch! {take1;
        TokenKind::CurrentPropertyGraph | TokenKind::CurrentGraph => empty.value(GraphExpr::Current),
        _ => fail
    }
    .spanned()
    .parse_next(input)
}

pub fn object_expression_primary(input: &mut TokenStream) -> ModalResult<Spanned<ObjectExpr>> {
    dispatch!{peek1;
        TokenKind::Variable => preceded(TokenKind::Variable, value_expression_primary).map(ObjectExpr::Variable),
        TokenKind::LeftParen => parenthesized_value_expression.map(ObjectExpr::Expr),
        _ => non_parenthesized_value_expression_primary_special_case.map(ObjectExpr::Expr),
    }.spanned().parse_next(input)
}

#[cfg(all(test, feature = "serde"))]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;
    use crate::parser::utils::parse;

    #[test]
    fn test_graph_expression_1() {
        let parsed = parse!(graph_expression, "current_property_graph");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_expression_2() {
        let parsed = parse!(graph_expression, "home_graph");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_expression_3() {
        let parsed = parse!(graph_expression, "/a/b");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_expression_4() {
        let parsed = parse!(graph_expression, "g");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_graph_expression_5() {
        let parsed = parse!(graph_expression, "variable a");
        assert_yaml_snapshot!(parsed);
    }
}
