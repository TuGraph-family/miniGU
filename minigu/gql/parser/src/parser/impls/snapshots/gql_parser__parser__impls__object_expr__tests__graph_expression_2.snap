---
source: minigu/parser/src/parser/impls/object_expr.rs
expression: parsed
---
- Ref: Home
- start: 0
  end: 10
